platform/linux-generic/include/odp_chksum_internal.h

```diff
@@ -0,0 +1,59 @@
+/* Copyright (c) 2013, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/**
+ * @file
+ *
+ * ODP checksum - implementation internal
+ */
+
+#ifndef ODP_CHKSUM_INTERNAL_H_
+#define ODP_CHKSUM_INTERNAL_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Simple implementation of ones complement sum.
+ * Based on RFC1071 and its errata.
+ */
+static uint32_t _odp_chksum_ones_comp16_32(const void *p, uint32_t len,
+					   odp_bool_t odd_offset)
```

Comment:
The `odd_offset` parameter is unnecessary here. What matters is whether `p` starts on an even or odd byte boundary since that determines whether you can fetch `data` as `uint16_t` quantities without making unaligned storage references. Each call to `_odp_chksum_ones_comp16_32()` is independent in this regard, so it must always be determined anew for each call.

Reply:
It is an offset insite packet.

Reply:
sorry. odd_offset references to an offset inside the packet. If you start calculating checksum from odd offset, you have to insert additional 0.

Reply:
The calculation starts from the address passed in as `p`. Whether that address is odd or even doesn't need a separate parameter to determine. The various ODP packet routines that return packet addresses, with the exception of `odp_packet_align()`, make no guarantees as to the alignment of any address they return.

Reply:
Calculation starts at address p, but an offset inside packet might be odd or even. If it is odd, we need to accomodate. Consider the code that uses it.

Reply:
What offset? The calculation is for `len` bytes starting at address `p`. Whether `p` is the address of an even or odd offset in its containing packet is irrelevant since what matters is whether `p` itself is even or odd, which is an independent question.

Reply:
Calculation of checksum starts at `l4_offset`. If for some reason this segment has odd amount of bytes after `l4_offset`, then checksumming will use that byte as `{b, 0}` `uint16_t` value. However calculation of checksum for the next segment (with first bytes `{c, d, e...}`) can not start as usual (`{c ,d}`, `{e, ....}). It should start as `{0, c}`, `{d, e}`, .... This is an odd offset. An offset in checksum calculation bytestream. Feel free to propose better name.

Reply:
OK, I see what you're saying now. But it looks like we need to consider both. The `odd_offset` parameter is saying whether the (continued) checksum needs to insert a leading zero into the calculation, however it's still true that you have to take into account whether `p` is even or odd since that can also vary independently. You can't fetch a `uint16_t` value from an odd byte address unless your CPU supports unaligned data accesses. Even for those machines that do, this typically results in a significant performance penalty.

But now we have another problem. If at entry `p` is even and `odd_offset` is true, meaning that the previous segment contained an odd number of bytes, then assuming this buffer is `{c, d, e, ...}` and we've added in `{0, c}` to account for the `odd_offset`, the next 16-bit value we want to sum is `{d, e}`, but `d` is on an odd address and `e` is on an even address so we can't fetch them as a `uint16_t` but need to assemble them from two `uint8_t` values. So it sounds like we need to factor out the logic required to fetch the next 16 bits to sum from the main loop:
```
static inline uint16_t data16(void *p) 
{
        if ((uintptr_t)p % 2)
                return (*((uint8_t *)p) << 8) + (*((uint8_t *)p + 1));

        return *((uint16_t *)p);
}
```
This can then be added to the main logic:
```
static uint32_t _odp_chksum_ones_comp_16_32(const void *p, uint32_t len, odp_bool_t odd_offset)
{
        void  *data = p;
        uint32_t sum = 0;
    
        if (odd_offset) {
               sum = *((uint8_t *)data++);
               len--;
        }

        while (len > 1) {
                sum += data16(data);
                data += 2;
                len -= 2;
        }
        
       ...etc.
}
```

Reply:
Hmm. Should we create a list of all possible unaligned access locations?

Reply:
One of the reasons why `odp_packet_copy_to_mem()` exists is to handle this sort of alignment problem since the caller can ensure that the destination memory address is aligned properly for intended use. But whenever you use `odp_packet_offset()` or similar routines there's no guarantee that the address you get back will have any particular alignment associated with it. Packets may start out aligned, but once head push/pull operations are issued all bets are off. 

Applications are responsible for keeping track of this sort of thing themselves, but for ODP internal use we have to be prepared for packets that have been manipulated in such a way that individual byte offsets may not have any particular alignment.

Reply:
It appears that the comments here still apply and need to be addressed.